File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/dc37cd46e578d29248104389008598aa.py
import sys
from typing import List, Tuple

class Fenw:
    def __init__(self):
        self.tr = [0] * 202020
    
    def add(self, pos: int) -> None:
        for i in range(pos, 200020, i | (i + 1)):
            self.tr[i] += 1
    
    def sum(self, pos: int) -> int:
        res = 0
        for i in range(pos, -1, (i & (i + 1)) - 1):
            res += self.tr[i]
        return res
    
    def range_sum(self, l: int, r: int) -> int:
        return self.sum(r) - self.sum(l - 1)

def solve1(v: List[Tuple[int, int]]) -> int:
    allF = []
    for p in v:
        from_ = (p[0], p[1] + w)
        to = (p[0], p[1] - w)
        allF.append(from_)
        allF.append(to)
    allF.append((0, 1))
    allF.sort()
    allF = list(set(allF))
    for i in range(len(v)):
        from_ = (v[i][0], v[i][1] + w)
        to = (v[i][0], v[i][1] - w)
        v[i] = (allF.index(from_), allF.index(to))
    v.sort(key=lambda x: (x[0], x[1]))
    res = 0
    f1 = Fenw()
    for i in range(len(v)):
        res += i - f1.sum(v[i][1] - 1)
        f1.add(v[i][1])
    return res

def solve2(v: List[List[Tuple[int, int]]]) -> int:
    allF = []
    for i in range(2):
        for p in v[i]:
            from_ = (p[0], p[1] + w)
            to = (p[0], p[1] - w)
            allF.append(from_)
            allF.append(to)
    allF.append((0, 1))
    allF.sort()
    allF = list(set(allF))
    res = 0
    f = [Fenw() for _ in range(2)]
    for i in range(2):
        for p in v[i]:
            from_ = (p[0], p[1] + w)
            to = (p[0], p[1] - w)
            p = (allF.index(from_), allF.index(to))
            f[i].add(p[0])
    for p in v[0]:
        res += f[1].range_sum(p[0], p[1])
    for p in v[1]:
        res += f[0].range_sum(p[0] + 1, p[1])
    return res

def solve() -> int:
    n = int(input())
    w = int(input())
    x = [0] * n
    a = [0] * n
    v = [[] for _ in range(2)]
    for i in range(n):
        x[i], a[i] = map(int, input().split())
        v[x[i] < 0].append((abs(x[i]), abs(a[i])))
    ans = 0
    ans += solve1(v[0])
    ans += solve1(v[1])
    ans += solve2(v)
    return ans

if __name__ == "__main__":
    print(solve())

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/dc37cd46e578d29248104389008598aa.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int p;
    int q;
} F;

typedef struct {
    int tr[202020];
} Fenw;

int x[101010];
int a[101010];
Fenw f1;
Fenw f[2];

int min(int a, int b) {
    return (a < b) ? a : b;
}

int max(int a, int b) {
    return (a > b) ? a : b;
}

void add(Fenw *fenw, int pos) {
    for (int i = pos; i <= 200020; i |= i + 1) {
        fenw->tr[i]++;
    }
}

int sum(Fenw *fenw, int pos) {
    int res = 0;
    for (int i = pos; i >= 0; i = (i & (i + 1)) - 1) {
        res += fenw->tr[i];
    }
    return res;
}

int solve1(int v[][2], int size, int w) {
    F allF[2 * size + 1];
    int allFSize = 0;
    for (int i = 0; i < size; i++) {
        F from = {v[i][0], v[i][1] + w};
        F to = {v[i][0], v[i][1] - w};
        allF[allFSize++] = from;
        allF[allFSize++] = to;
    }
    allF[allFSize++] = (F){0, 1};
    qsort(allF, allFSize, sizeof(F), [](const void *a, const void *b) {
        F *f1 = (F *)a;
        F *f2 = (F *)b;
        if (f1->p * 1LL * f2->q < f2->p * 1LL * f1->q) {
            return -1;
        } else if (f1->p * 1LL * f2->q > f2->p * 1LL * f1->q) {
            return 1;
        } else {
            return 0;
        }
    });
    int vSize = sizeof(v) / sizeof(v[0]);
    for (int i = 0; i < vSize; i++) {
        F from = {v[i][0], v[i][1] + w};
        F to = {v[i][0], v[i][1] - w};
        int fromIndex = -1;
        int toIndex = -1;
        for (int j = 0; j < allFSize; j++) {
            if (from.p == allF[j].p && from.q == allF[j].q) {
                fromIndex = j;
            }
            if (to.p == allF[j].p && to.q == allF[j].q) {
                toIndex = j;
            }
        }
        v[i][0] = fromIndex;
        v[i][1] = toIndex;
    }
    qsort(v, vSize, sizeof(v[0]), [](const void *a, const void *b) {
        int *v1 = (int *)a;
        int *v2 = (int *)b;
        if (v1[0] < v2[0] || (v1[0] == v2[0] && v1[1] > v2[1])) {
            return -1;
        } else if (v1[0] > v2[0] || (v1[0] == v2[0] && v1[1] < v2[1])) {
            return 1;
        } else {
            return 0;
        }
    });
    long long res = 0;
    memset(f1.tr, 0, sizeof(f1.tr));
    for (int i = 0; i < vSize; i++) {
        res += i - sum(&f1, v[i][1] - 1);
        add(&f1, v[i][1]);
    }
    return res;
}

long long solve2(int v[][2], int size, int w) {
    F allF[2 * size + 1];
    int allFSize = 0;
    for (int i = 0; i < size; i++) {
        F from = {v[i][0], v[i][1] + w};
        F to = {v[i][0], v[i][1] - w};
        allF[allFSize++] = from;
        allF[allFSize++] = to;
    }
    allF[allFSize++] = (F){0, 1};
    qsort(allF, allFSize, sizeof(F), [](const void *a, const void *b) {
        F *f1 = (F *)a;
        F *f2 = (F *)b;
        if (f1->p * 1LL * f2->q < f2->p * 1LL * f1->q) {
            return -1;
        } else if (f1->p * 1LL * f2->q > f2->p * 1LL * f1->q) {
            return 1;
        } else {
            return 0;
        }
    });
    int vSize = sizeof(v) / sizeof(v[0]);
    for (int i = 0; i < vSize; i++) {
        F from = {v[i][0], v[i][1] + w};
        F to = {v[i][0], v[i][1] - w};
        int fromIndex = -1;
        int toIndex = -1;
        for (int j = 0; j < allFSize; j++) {
            if (from.p == allF[j].p && from.q == allF[j].q) {
                fromIndex = j;
            }
            if (to.p == allF[j].p && to.q == allF[j].q) {
                toIndex = j;
            }
        }
        v[i][0] = fromIndex;
        v[i][1] = toIndex;
        add(&f[i % 2], v[i][0]);
    }
    long long res = 0;
    for (int i = 0; i < vSize; i++) {
        res += sum(&f[(i + 1) % 2], v[i][0], v[i][1]);
    }
    return res;
}

long long solve(int n, int w) {
    int v[2][101010][2];
    int vSize[2] = {0, 0};
    for (int i = 0; i < n; i++) {
        scanf("%d %d", &x[i], &a[i]);
        if (x[i] < 0) {
            v[0][vSize[0]][0] = abs(x[i]);
            v[0][vSize[0]][1] = abs(a[i]);
            vSize[0]++;
        } else {
            v[1][vSize[1]][0] = abs(x[i]);
            v[1][vSize[1]][1] = abs(a[i]);
            vSize[1]++;
        }
    }
    long long ans = 0;
    ans += solve1(v[0], vSize[0], w);
    ans += solve1(v[1], vSize[1], w);
    ans += solve2(v[0], vSize[0], w);
    ans += solve2(v[1], vSize[1], w);
    return ans;
}

int main() {
    int n, w;
    scanf("%d %d", &n, &w);
    printf("%lld\n", solve(n, w));
    return 0;
}
